patches/api/0332-More-Teleport-API.patch

```diff
@@ -0,0 +1,175 @@
++     * <p>
++     * Note: You cannot keep passengers if you are teleporting this entity to another world.
++     *
++     * @param location New location to teleport this entity to
++     * @param keepPassengers If passengers should be kept when teleporting
++     * @return <code>true</code> if the teleport was successful
++     */
++    public boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean keepPassengers);
++
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
++     *
++     * @param location New location to teleport this entity to
++     * @param keepPassengers If passengers should be kept when teleporting
++     * @param dismount If the entity should be dismounted if they are riding another entity
++     * @return <code>true</code> if the teleport was successful
++     */
++    public boolean teleport(@NotNull Location location, boolean keepPassengers, boolean dismount);
++
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
++     *
++     * @param location New location to teleport this entity to
++     * @param keepPassengers If passengers should be kept when teleporting
++     * @param dismount If the entity should be dismounted if they are riding another entity
++     * @return <code>true</code> if the teleport was successful
++     */
++    public boolean teleport(@NotNull Location location, @NotNull TeleportCause cause, boolean keepPassengers, boolean dismount);
++    // Paper end - Teleport API
++
+     /**
+      * Teleports this entity to the given location. If this entity is riding a
+      * vehicle, it will be dismounted prior to teleportation.
+diff --git a/src/main/java/org/bukkit/entity/Player.java b/src/main/java/org/bukkit/entity/Player.java
+index afa245018192a63e5db8bc568ddea2790bf5b8a2..e8a7a0477db682292de043256272ba0f2f9f0b78 100644
+--- a/src/main/java/org/bukkit/entity/Player.java
++++ b/src/main/java/org/bukkit/entity/Player.java
+@@ -2289,6 +2289,52 @@ public interface Player extends HumanEntity, Conversable, OfflinePlayer, PluginM
+     String getClientBrandName();
+     // Paper end
+ 
++    // Paper start - Teleport API
++    /**
++     * Teleports this entity to the given location.
++     * <p>
++     * Note: You cannot keep passengers or stay mounted if you are teleporting this entity to another world.
++     *
++     * <p>
++     * Relative teleportation flags are only used client side, and cause the player to not lose velocity in that
++     * specific coordinate. The location of the teleportation will not change.
++     *
++     * @param location New location to teleport this entity to
++     * @param keepPassengers If passengers should be kept when teleporting
```

Comment:
The description here is incorrect; it's not about whether they should be kept, but whether teleporting should be possible *at all* with passengers (same for everywhere else)